core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp

```diff
@@ -820,27 +820,83 @@ class alignas(FloatType) floating_point_wrapper {
     return tmp_lhs < tmp_rhs;
   }
 
+  template <class T>
```

Comment:
Why template functions and not non-template overloads?

I'm asking because they don't do the same thing (such as how implicit conversions are handled).

Reply:
the template versions in principle allow us to group things together (here float and double), we can do the same thing for the integer versions in another PR where we would have to cast the half precision thing to float instead of T. From an implicit conversion thing for the common case this just changes from "no-overload found" to "ambiguous overload" or? Say we had a float and a double non-template overlaod and someone would compare with int?

Reply:
> Why template functions and not non-template overloads?

I agree with @crtrott and would add that it reduces code duplication. But, I may be missing something. I do not see how this mixing member and non-member overloads of operatorOP would affect overload resolution. The compiler either:
1. has a conflict from the implicit conversion constructor and the member overload of operatorOP as well as the non-member overload for float and double or;
2. does not find a overload of operatorOP or;
3. finds the correct overload of operatorOP.

From https://en.cppreference.com/w/cpp/language/overload_resolution:

>  Call to an overloaded operator
>
> If at least one of the arguments to an operator in an expression has a class type or an enumeration type, both [builtin operators](https://en.cppreference.com/w/cpp/language/expressions#Operators) and [user-defined operator overloads](https://en.cppreference.com/w/cpp/language/operators) participate in overload resolution, with the set of candidate functions selected as follows:
>
> For a unary operator @ whose argument has type T1 (after removing cv-qualifications), or binary operator @ whose left operand has type T1 and right operand of type T2 (after removing cv-qualifications), the following sets of candidate functions are prepared:
> 1) member candidates: if T1 is a complete class or a class currently being defined, the set of member candidates is the result of [qualified name lookup](https://en.cppreference.com/w/cpp/language/lookup) of T1::operator@. In all other cases, the set of member candidates is empty.
> 2) non-member candidates: For the operators where [operator overloading](https://en.cppreference.com/w/cpp/language/operators) permits non-member forms, all declarations found by [unqualified name lookup](https://en.cppreference.com/w/cpp/language/lookup) of operator@ in the context of the expression (which may involve [ADL](https://en.cppreference.com/w/cpp/language/adl)), except that member function declarations are ignored and do not prevent the lookup from continuing into the next enclosing scope. If both operands of a binary operator or the only operand of a unary operator has enumeration type, the only functions from the lookup set that become non-member candidates are the ones whose parameter has that enumeration type (or reference to that enumeration type)
> 3) built-in candidates: For operator,, the unary operator&, and operator->, the set of built-in candidates is empty. For other operators built-in candidates are the ones listed in [built-in operator pages](https://en.cppreference.com/w/cpp/language/expressions#Operators) as long as all operands can be implicitly converted to their parameters. If any built-in candidate has the same parameter list as a non-member candidate or rewritten non-member candidate (since C++20) that is not a function template specialization, it is not added to the list of built-in candidates. When the built-in assignment operators are considered, the conversions from their left-hand arguments are restricted: user-defined conversions are not considered.
> 4) rewritten candidates:
>        For the four relational operator expressions x < y, x <= y, x > y, and x >= y, all member, non-member, and built-in operator<=>s found are added to the set.
 >       For the four relational operator expressions x < y, x <= y, x > y, and x >= y as well as the three-way comparison expression x <=> y, a synthesized candidate with the order of the two parameters reversed is added for each member, non-member, and built-in operator<=>s found.
 >       For x != y, all member, non-member, and built-in operator==s found are added to the set, unless there is a matching operator!=.
 >       For equality operator expressions x == y and x != y, a synthesized candidate with the order of the two parameters reversed is added for each member, non-member, and built-in operator==s found, unless there is a matching operator!=.
> In all cases, rewritten candidates are not considered in the context of the rewritten expression. For all other operators, the rewritten candidate set is empty. 
> ...
> User-defined conversions (both converting constructors and user-defined conversion functions) are prohibited from taking part in implicit conversion sequence where it would make it possible to apply more than one user-defined conversion. Specifically, they are not considered if the target of the conversion is the first parameter of a constructor or the implicit object parameter of a user-defined conversion function, and that constructor/user-defined conversion is a candidate for 
> ...
> Additional rules for function template candidates
> If name lookup produces a function template, [template argument deduction](https://en.cppreference.com/w/cpp/language/template_argument_deduction) and checking of any explicit template arguments are performed to find the template argument values (if any) that can be used in this case:
>    if both succeeds, the template arguments are used to synthesize declarations of the corresponding function template specializations, which are added to the candidate set, and such specializations are treated just like non-template functions except where specified otherwise in the tie-breaker rules below;
>    if argument deduction fails or the synthesized function template specialization would be ill-formed, no such function is added to the candidate set (see [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae)). 
> If a name refers to one or more function templates and also to a set of overloaded non-template functions, those functions and the specializations generated from the templates are all candidates.

Reply:
Given

```c++
struct MyDouble {
    operator double() const { /* ... */ }
};
```

You could not pass `MyDouble` to it, so this doesn't act like built in comparison operators.

In this file, we have some comparisons that are member functions, some are non-templated `friend` functions, and now some that are templated `friend` functions, as well as sometimes passing by value to avoid `const`/`volatile` issues.  That combination is very hard to reason about.

Reply:
```c++
struct MyDouble {
    operator double() const { /* ... */ }
};
```

You could not pass `MyDouble` to it, so this doesn't act like built in comparison operators.

In this file, we have some comparisons that are member functions, some are non-templated `friend` functions, and now some that are templated `friend` functions, as well as sometimes passing by value to avoid `const`/`volatile` issues.  That combination is very hard to reason about.

Reply:
@nliber I see. We can replace all the templated functions with non-templated functions. But, AFAIK, we cannot replace the non-member friend functions for all binary operators since they are used to overload the LHS type.

@crtrott, @dalg24, @ajpowelsnl: With half_t and bhalf_t in a minimally working state, I suggest we take a step back at this point and document a strategy for how to 'make half_t and bhalf_t behave like a c++ built-in type'. I am assuming this will not be easy given that the rules surrounding this are quite complicated.

Reply:
@e10harvey the only operators that are required to be member functions are assignment `=`, subscript `[]`, call `()` and pointer indirection `->`.

For all the operators that don't take a non-const reference to the LHS, I'd rather they all be hidden `friend`s, which matches the current wisdom of the C++ standard library.  Newer libraries in the C++ standard library follow this convention.  See [P1601](https://wg21.link/P1601).

Reply:
Got it. Thank you for the detailed review and guidance, @nliber.

Shall we make this change (everything but `=` operator implemented as a hidden friend) in this bug fix PR or in a follow-on issue/PR.

Reply:
CC: @crtrott, @dalg24

Reply:
The failing HIP Jenkins CI SIGINT failure appears to be unrelated to these changes.

Reply:
@e10harvey A follow-on is fine by me.

Reply:
@nliber See https://github.com/kokkos/kokkos/issues/6468.